### ISSUE TITLE
Content-Security-Policy prevents iframing

### ISSUE THREAD

**Issue by ltalirz** [materialscloud-org/tools-optimade-client#12]
I'm not sure which repository is the right place to post this, but for some reason, the optimadeclient tool is sending `Content-Security-Policy: frame-ancestors 'self'; report-uri /optimadeclient/api/security/csp-report` headers:

(note: I'm checking on dev-tools to be sure this has nothing to do with dokku)
```
$ curl -X GET -H "Origin: http://www.materialscloud.org" -I https://dev-tools.materialscloud.org/optimadeclient/
HTTP/1.1 200 OK
Date: Fri, 16 Oct 2020 00:14:31 GMT
Server: TornadoServer/6.0.3
Content-Type: text/html
Content-Security-Policy: frame-ancestors 'self'; report-uri /optimadeclient/api/security/csp-report
Vary: Accept-Encoding
Transfer-Encoding: chunked
```
This prevents iframing on materialscloud.org.

For comparison, these are the headers sent by e.g. seekpath:
```
$ curl -X GET -H "Origin: http://www.materialscloud.org" -I https://dev-tools.materialscloud.org/seekpath/
HTTP/1.1 200 OK
Date: Fri, 16 Oct 2020 00:17:12 GMT
Server: Apache/2.4.29 (Ubuntu)
Content-Length: 14528
Vary: Accept-Encoding
Content-Type: text/html; charset=utf-8
```

**Comment by ltalirz** [comment]
The tornado settings can be configured like this: https://github.com/voila-dashboards/voila/pull/69#issue-254770444

It's not quite clear to me who sets the default here - tornado, voila, the template, the app ...

**Comment by CasperWA** [comment]
Right! This issue belongs in the materialscloud-org/voila-materialscloud-template repository - unless we apply the fix directly here. I'll try to first fix it in the template, if that doesn't work, I'll go the "per use case" route.

**Comment by CasperWA** [comment]
Hmm, thinking more about it, it doesn't make sense to be able to change this from the template, hence it _has to be_ "per use case", i.e., every iframed Voilà application needs to run the tornado server using this specific configuration for it work with iframing.

**Comment by CasperWA** [comment]
I have spent an unseemly amount of time on figuring this out when in the dokku environment. Now, for whatever reason, it suddenly worked, and I didn't anymore get the error

```sh
setuidgid: fatal: unable to run ./run.sh: file does not exist
```

I think it may be related to the issue described [here](https://github.com/dokku/dokku/issues/1916#issuecomment-247757417), but it's difficult to figure out.

In any case, I now get the proper header result for http://optimadeclient.matcloud.xyz:

```sh
$ curl -X GET -H "Origin: https://www.materialscloud.org" -I http://optimadeclient.matcloud.xyz
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 16 Oct 2020 11:36:07 GMT
Content-Type: text/html
Transfer-Encoding: chunked
Connection: keep-alive
Vary: Accept-Encoding
Content-Security-Policy: frame-ancestors '*' localhost:*; report-uri /api/security/csp-report
```

**Comment by CasperWA** [comment]
I have pushed this change also to https://dev-tools.materialscloud.org/optimadeclient/, although here I'm not using the `materialscloud-iframe` template, but the `materialscloud` template, i.e., if iframed it will have two headers.
I.e., @ltalirz if you're testing this, try to maybe use http://optimadeclient.matcloud.xyz instead :)

**Comment by ltalirz** [comment]
I've updated the materialscloud development server: https://dev-www.materialscloud.org/work/tools/optimadeclient

Note: I'm iframing the production dokku instance, since I'll need to wait for giovanni to be able to get an SSL certificate for the development instance as well (and since dev-www uses https, it won't like iframing tools over http from matcloud.xyz).

For the moment, I've made the CSP more open again, since I found that your `frame-ancestors https://*.materialscloud.org localhost:*` was blocking iframing for my local testing of the tools frontend
(I don't know why - the URL was http://localhost:8000/tools/optimadeclient ).
If you like you can play around more with the frame-ancestors but for the moment I think preventing other people from iframing our tools is not our main concern.

From my point of view, there are a few remaining important issues:
 * (dokku) the lack of the loading banner - at some point we will need figure out why it does not show on dokku; very weird
 * (voila) whether we can make the CSP configuration easier. Moving stuff to a separate run.sh is something I would like to avoid and, to be honest, if we can't incorporate the setting into the materials cloud template, we may even consider not iframing voila tools by default - it seems a significant overhead and potential for confusion/friction with app developers for little gain
 * (optimadeclient) it may be possible to improve the load time of the tool quite a bit. it's voila, so it will never be lightning fast but other tools load significantly faster, e.g. http://periodictable.matcloud.xyz/

**Comment by CasperWA** [comment]
Didn't know I pushed the iframed version to the production server of dokku actually.
And weird it didn't work with your local testing. Perhaps it's the mixing of specifying an ancestor with `https` and not?
I could try with just `*.matierialscloud.org` instead?

Concerning the loading time, it performs some initial requests, which should already be significantly improved with the newly implemented caching. However, I have noticed 3-4 seconds are spent not due to the notebook but the template for this particular notebook. I think part of the issue there is we need to clean up the template as well (see materialscloud-org/voila-materialscloud-template#17.
But I also have an open issue to perform the requests asynchronously (see CasperWA/voila-optimade-client#138).

Otherwise, the issue pertaining to iframing Voilà applications is indeed a serious one. We can perhaps mediate it slightly if we setup the automatic deployment of Voilà applications similar to what is done for binder, where a `jupyter_config.json` file in the repository root is copied to Jupyter's configuration directory locally and renamed to `voila.json`. This could automatically set the header and more, and the JSON is easy to handle for developers, I hope?
